include/rdma/providers/fi_peer.h

```diff
@@ -164,6 +164,8 @@ struct fi_ops_srx_owner {
 			size_t size, uint64_t tag, struct fi_peer_rx_entry **entry);
 	int	(*queue_msg)(struct fi_peer_rx_entry *entry);
 	int	(*queue_tag)(struct fi_peer_rx_entry *entry);
+	int	(*queue_msg_progress)(struct fi_peer_rx_entry *entry);
+	int	(*queue_tag_progress)(struct fi_peer_rx_entry *entry);
```

Comment:
The only know issue with the current APIs is that the address may not be known until after the message is queued.  That leads me to consider a callback such as ```(*modify_addr)(struct fi_peer_rx_entry *entry, fi_addr_t addr)```.

For the peer to make such a call, it would need to track fi_peer_rx_entry when the address isn't known.  I don't like the idea that both the peer and owner would need to maintain references to fi_peer_rx_entry.  Passing the structure back and forth seems far more desirable for managing the lifetime.  Soooo....

I'm wondering if we shouldn't reverse the flow here.  As proposed, the owner will iterate over all queued messages that don't have an address, so that the peer can update the entry.  We could instead place the burden on the peer to notify the owner that an address has changed.  This could look similar to a list iterator operation.  For example,

```
owner::int (*update_addr)(struct fid_peer_srx *srx, fi_addr_t (*get_addr)(struct fi_peer_rx_entry *entry));
```
1. The peer calls update_addr() (Possibly in response to an AV insertion.  We could instead require the callback be deferred until normal CQ progress is called for easier synchronization.)
2. The owner iterates over the rx_entry list for that peer, where the src_addr is set to NOTAVAIL.
3. The owner calls the get_addr() method for each rx_entry
4. The peer either returns an updated fi_addr_t value or NOTAVAIL if it's still unknown.

If the peer isn't configured with FI_DIRECTED_RECV, then the address can be set to anything.  But the flow of calling get_msg() + queue_msg() remains the same.  Otherwise, when FI_DIRECTED_RECV is not set, which call flow to use isn't obvious.

Reply:
This is the current behavior for tcp:

If a message arrives that is unmatched and the peer address is unknown, the message is left on the socket and socket events are disabled.  The unexpected queue is size 1 until it knows the peer's address.  When a new address is inserted into the AV, a 'peer' data structure is updated with the address, but there's no message progress made.  Basically, there's nothing that can be done with this socket until a receive is posted.

When a receive is posted, it starts progress.  If the receive is for ADDR_UNSPEC, all sockets with a message queued on it are checked for a possible match.  If the receive targets a specific src, the same thing happens.  However, at this point, we know that the src address has been inserted into the AV, and the peer structure would have been updated with an address.  So, the src address check is part of the matching.

In both cases, the queued message is received into the posted buffer.  Once done, the process repeats.  If a message arrives that is unmatched and the peer address is unknown, the message is left on the socket... blah blah.

Sooo... I think tcp can be updated to handle an update_addr() flow.  Instead iterating over the list of sockets with unmatched messages when a receive is posted, AV insert iterates over the list, finds which socket has been updated (if any), and updates the src_addr.  Note that a call to update_addr() can never result in matching a posted receive buffer and would be invoked outside of any progress loop.

Reply:
As long as you're sure tcp can be updated to move that to the AV instead of the recv path, that's fine with me. I just wasn't sure.
So peer always calls get/queue but has to keep track of all unspec entries on its own in order to call update_addr later?
owner can check the unspec addr list for an unspec addr recv post
on peer av insert, peer goes through the list of rx entries for the updated peer and calls owner_srx->update_addr on each one?
on start_msg, if the rx_entry/unexpected message is unspec addr, the peer is also responsible for removing the entry from its queued update list so it doesn't call update addr on it again?

the part about the peer being responsible for duplicate tracking it the part I want to verify. This is going to add more complexity on the peer's side

Reply:
I'm fairly sure tcp can move to this model.

The peer does *not* need to track entries.  Example code:

```
/* peer */
fi_addr_t peer_get_addr(struct fi_peer_rx_entry *entry)
{
    dest_context = entry->peer_context;
    return dest_ip_array[dest_context->dest_ip_index].fi_addr
}

peer_av_insert(..dest_ip..)
{
    av[next].dest_ip = dest_ip
    x = find index of dest_ip in dest_ip_array[]
    if (x == -1) {
        dest_ip_array[last++].fi_addr = next++;
        return;
    }

    assert(dest_ip_array[x].fi_addr == FI_ADDR_UNSPEC)
    dest_ip_array[x].fi_addr = next++;
    srx->owner->update_addr(srx, peer_get_addr)
}
```

```
/* owner */
owner_update_addr(..., get_addr())
{
    for each entry in unknown addr list {
        fi_addr =get_addr(entry)
        if (fi_addr != FI_ADDR_UNSPEC) {
            remove entry
            insert entry on src_addr_list
        }
    }
}
```

1. address is inserted into peer AV
2. peer always calls srx->update_addr()
3. owner iterates over all saved entries for that peer where the address is UNSPEC.
3a. owner calls peer->get_addr()
3b. peer returns the new address for the entry or UNSPEC.
3c. owner updates the saved entry

The peer needs some way to know which source a message arrived on, so that once the source is inserted into the AV, it can update it correctly.

Note: get_addr() could be set earlier, but I defined the API as shown to make the operation clearer, which apparently didn't work.

Reply:
Ah, that makes more sense. I thought you were just thinking of a single update_addr call that the peer would call into the owner for a single entry. This makes more sense with the get_addr call as well. Thanks!